pkg/cloudprovider/cloudprovider.go

```diff
@@ -13,17 +16,70 @@ const (
 )
 
 var (
+	instance             Ops
+	once                 sync.Once
 	providerRegistry     map[string]Ops
 	providerRegistryLock sync.Mutex
 )
 
 // Ops is a list of APIs to fetch information about cloudprovider and its nodes
 type Ops interface {
+	// init checks environment and initializes cloud provider
+	init(nodeList *v1.NodeList) error
+
 	// Name returns the name of the cloud provider
 	Name() string
 
 	// GetZone returns the zone of the provided node
 	GetZone(*v1.Node) (string, error)
+
+	// IsK8sServiceProvider checks whether it's running kubernetes service provided by the same provider
+	IsK8sServiceProvider() bool
+}
+
+// Instance returns a singleton instance of cloud provider
+func Instance() Ops {
+	once.Do(func() {
+		if instance == nil {
+			instance = New("")
+		}
+	})
+	return instance
+}
+
+// SetInstance replaces the instance with the provided one. Should be used only for testing purposes.
+func SetInstance(i Ops) {
+	instance = i
+}
+
+// InitCloudProvider initialize the cloud provider instance at operator starts
+func InitCloudProvider() error {
+	initRegistry()
+
+	nodeList, err := coreops.Instance().GetNodes()
+	if err != nil {
+		return err
+	}
+
+	var cloudProviderName string
+	for _, node := range nodeList.Items {
+		// Get the cloud provider
+		// From kubernetes node spec: <ProviderName>://<ProviderSpecificNodeID>
+		if node.Spec.ProviderID != "" {
+			tokens := strings.Split(node.Spec.ProviderID, "://")
+			if len(tokens) == 2 {
+				cloudProviderName = tokens[0]
+				break
+			} // else provider id is invalid
+		}
+	}
+
+	instance = New(cloudProviderName)
```

Comment:
We should add a check if the cloud provider name never gets set? Or is this handled inside the `New()` method?

Reply:
if it's not a registered provider, the name can be registered as an empty provider. it's used somewhere else as well, don't want to break things.